### PR TITLE
fix(link): change package version when exec link

### PR DESCRIPTION
**Pull request by BlackHole1**
Fixed: #4341

I will add the relevant unit tests tomorrow.
Before that, I need @zkochan your code review. I am not sure if this is a reasonable change.

### CODE REVIEW

File: packages/core/src/link/index.ts [hunk 1]
```diff
@@ -78,10 +78,13 @@ export default async function link (
       throw new PnpmError('INVALID_PACKAGE_NAME', `Package in ${linkFromPath} must have a name field to be linked`)
     }
 
+    // if the package already exists and lock version, its version should not be changed
+    const shouldModifyVersion = maybeOpts.linkToBin ?? !/^\d.*/.test((opts.manifest?.dependencies ?? {})[manifest.name])
```
Comment (BlackHole1) [inline packages/core/src/link/index.ts]: `maybeOpts.linkToBin`: filter `pnpm link --global`

File: packages/core/src/link/index.ts [hunk 2]
```diff
@@ -78,10 +78,13 @@ export default async function link (
...
+
     specsToUpsert.push({
       alias: manifest.name,
       pref: getPref(manifest.name, manifest.name, manifest.version, {
-        pinnedVersion: opts.pinnedVersion,
+        pinnedVersion: shouldModifyVersion ? opts.pinnedVersion : 'patch',
```
Comment (BlackHole1) [inline packages/core/src/link/index.ts]: `patch` will not change version

**Comment by BlackHole1** [comment]
I found it difficult to write unit tests for this change.
@zkochan  Do you have any ideas?

**Comment by zkochan** [comment]
It will be solved by this PR: https://github.com/pnpm/pnpm/pull/6513

**Pull request closed without merging**